96.md

```diff
@@ -0,0 +1,325 @@
+
+`Clients` must add a [NIP-98](98.md) `Authorization` header with the encoded `payload` tag set to the base64-encoded 256-bit SHA-256 hash of the file (not the hash of the whole request body).
+If using an html form, use an `Authorization` form data field instead.
+
+These following **optional** form data fields MAY be used by `servers` and SHOULD be sent by `clients`:
+- `expiration`: string of the UNIX timestamp in seconds. Empty string if file should be stored forever. The server isn't required to honor this;
+- `size`: string of the file byte size. This is just a value the server can use to reject early if the file size exceeds the server limits;
+- `alt`: strict description text for visibility-impaired users;
+- `caption`: loose description.
+
+Others custom form data fields may be used depending on specific `server` support.
+The `server` isn't required to store any metadata sent by `clients`.
+
+Note for `clients`: if using an HTML form, it is important for the `file` form field to be the **last** one, or be re-ordered right before sending or be appended as the last field of XHR2's FormData object.
+
+The `filename` embedded in the file may not be honored by the `server`, which could internally store just the SHA-256 hash value as the file name, ignoring extra metadata.
+The hash is enough to uniquely identify a file, that's why it will be used on the "download" and "delete" routes.
+
+The `server` MUST link the user's `pubkey` string (which is embedded in the decoded header value) as the owner of the file so to later allow him to delete the file.
+Note that if a file with the same hash of a previously received file (so the same file) is uploaded by another user, the server doesn't need to store the new file.
+It should just add the new user's `pubkey` to the list of the owners of the already stored file with said hash.
+
+The `server` SHOULD also store the `Authorization` header/field value (decoded or not) for accountability purpose as this proves that the user with the unique pubkey did ask for the upload.
+
+The `server` MUST reject with 413 Payload Too Large if file size exceeds limits.
+
+The `server` MUST reject with 401 Unauthorized status if some fiels are invalid.
+
+The `server` MUST reject the upload with 403 Forbidden status if the `payload` tag value contains an already used SHA-256 hash (if file is already owned by the same pubkey) or it isn't the same of the received file.
+
+The `server` MAY reject the upload with 402 Payment Required status if the user has a pending payment (Although payment flow is not strictly required. Server owners decide if the storage is free or not. Although we suggest [this freemium monetization scheme](#monetization))
+
+The upload response is a json as follows:
+
+```js
+{
+  nip96: {
+    // SHA-256 hash. Empty string if unsuccessful
+    x: "719171db19525d9d08dd69cb716a18158a249b7b3b3ec4bbdec5698dca104b7b",
```

Comment:
It would be nice if the server could return a canonical URL for the hosted media. This would allow e.g. uploads to `api.example.com`, while serving the media from `cdn.example.com`. But the only way that works is if the server can tell the client where the media is accessed.

Reply:
The download is always at `api.example.com/nip96/the-x-value` both with or without a file extension

Reply:
Right. But imagine I have a CDN that I want to serve the files from that is separate from the API that accepts the uploads. If the server could tell the client the canonical URL for the media, you could facilitate this.

Reply:
yeah you should be able to have different hostnames

Reply:
@bndw @michaelhall923 Oh, sorry misunderstood. Now I get it.

However, what happens if the server using `cdn.example.com/...` for downloads in the future starts using another cdn and changes download urls to `anothercdn.example.com/...`? This wouldn't be a problem if the urls were used on a website you control but it does is a problem for urls used on nostr notes.

That's why in my opinion the best approach is for the server to simply issue a 302 redirect from the fixed download url to whatever cdn it wants.

Or what you want is a different hostname for the fixed url that then would issue the 302 redirects to the cdn? Like the upload (at `a-server.com/nip96`) response would have something like this:
```js
{
  nip96: {
    hostname: 'another.hostname.com' // not a cdn
    x: "719171db19525d9d08dd69cb716a18158a249b7b3b3ec4bbdec5698dca104b7b",
  },
  errors: {  nip96: [] }
}
```

Then you would be able to do a 302 redirect from `another.hostname.com/nip96/719171...` (instead of from `a-server.com/nip96/719171...`) to `cdn.example.com/nip96/719171...`?

The downside is some servers may wrongly reply with `{ nip96: {  hostname: 'cdn.example.com' } ... }` pointing directly to the cdn.

Reply:
Yeah the core of the idea is that GET requests would never hit the upload server. This has nice properties around availability and caching e.g. if the upload server (api.example.com) gets overwhelmed it does not have to impact serving files (cdn.example.com).

Reply:
This is now configurable at  /.well-known/nostr.json.
The response also has a "download_route" field (can add custom paths to, so no /nip96 path anymore)